airbyte-server/src/main/java/io/airbyte/server/handlers/helpers/CatalogConverter.java

```diff
@@ -111,20 +162,62 @@ public static io.airbyte.protocol.models.ConfiguredAirbyteCatalog toProtocolKeep
         .filter(s -> s.getConfig().getSelected())
-        .map(s -> new io.airbyte.protocol.models.ConfiguredAirbyteStream()
-            .withStream(toProtocol(s.getStream()))
-            .withSyncMode(Enums.convertTo(s.getConfig().getSyncMode(), io.airbyte.protocol.models.SyncMode.class))
-            .withCursorField(s.getConfig().getCursorField())
-            .withDestinationSyncMode(Enums.convertTo(s.getConfig().getDestinationSyncMode(),
-                io.airbyte.protocol.models.DestinationSyncMode.class))
-            .withPrimaryKey(Optional.ofNullable(s.getConfig().getPrimaryKey()).orElse(Collections.emptyList())))
+        .map(s -> {
+          try {
+            return new io.airbyte.protocol.models.ConfiguredAirbyteStream()
+                .withStream(toProtocol(s.getStream(), s.getConfig()))
+                .withSyncMode(Enums.convertTo(s.getConfig().getSyncMode(), io.airbyte.protocol.models.SyncMode.class))
+                .withCursorField(s.getConfig().getCursorField())
+                .withDestinationSyncMode(Enums.convertTo(s.getConfig().getDestinationSyncMode(),
+                    io.airbyte.protocol.models.DestinationSyncMode.class))
+                .withPrimaryKey(Optional.ofNullable(s.getConfig().getPrimaryKey()).orElse(Collections.emptyList()));
+          } catch (JsonValidationException e) {
+            LOGGER.error("Error parsing catalog: {}", e);
+            errors.add(e);
+            return null;
+          }
+        })
         .collect(Collectors.toList());
+    if (!errors.isEmpty()) {
+      throw errors.get(0);
+    }
     return new io.airbyte.protocol.models.ConfiguredAirbyteCatalog()
         .withStreams(streams);
   }
 
+  /**
+   * Generate the map from StreamDescriptor to indicator of whether field selection is enabled for
+   * that stream.
+   *
+   * @param syncCatalog the catalog
+   * @return the map as a FieldSelectionData object
+   */
+  public static FieldSelectionData getFieldSelectionData(AirbyteCatalog syncCatalog) {
+    if (syncCatalog == null) {
+      return null;
+    }
+    final var fieldSelectionData = new FieldSelectionData();
+    syncCatalog.getStreams().stream().forEach((streamAndConfig) -> {
+      final var streamDescriptor = new StreamDescriptor()
+          .name(streamAndConfig.getStream().getName())
+          .namespace(streamAndConfig.getStream().getNamespace());
+      final boolean fieldSelectionEnabled =
+          streamAndConfig.getConfig().getFieldSelectionEnabled() == null ? false : streamAndConfig.getConfig().getFieldSelectionEnabled();
+      fieldSelectionData.setAdditionalProperty(streamDescriptorToStringForFieldSelection(streamDescriptor), fieldSelectionEnabled);
+    });
+    return fieldSelectionData;
+  }
+
+  // Return a string representation of a stream descriptor that's convenient to use as a key for the
+  // field selection data.
+  private static String streamDescriptorToStringForFieldSelection(final StreamDescriptor streamDescriptor) {
```

Comment:
@alovew wanted to get your input here - As far as I can tell the `additionalProperties` construct needs to map `String` -> `Something`. I initially used the toString method but it's a bit clunky and I figured this has the same uniqueness and would be easier to deal with in logs/db queries/etc.

Is there some existing other standard on using StreamDescription like this?

Reply:
unfortunately I think this is implemented at least one other time (possibly more) in various places - probably something we should clean up at some point but I think it's ok if we go this way for now, but I'm not sure we have to:

I'm a little confused by the additional properties thing, I didn't notice this in the last PR. normally our objects look like this:

```
  schedule:
    type: object
    required:
      - timeUnit
      - units
    additionalProperties: false
    properties:
      timeUnit:
        type: string
        enum:
          - minutes
          - hours
          - days
          - weeks
          - months
      units:
        type: integer
```

but this one looks like this:
```
  fieldSelectionData:
    type: object
    description: A map of StreamDescriptor to an indicator of whether field selection is enabled for that stream.
    additionalProperties:
      type: boolean
```

I thought that to declare objects in json schema you had to have a properties field that nested the properties. and I think we can make a StreamDescriptor object one of the properties that is nested in the object

Reply:
This one is indeed a bit different, because it's mapping arbitrary keys to bool values. In other words if we declared a StreamDescriptor object as a property, that could just represent a single stream. We could do an array of (StreamDescriptor, bool)-pairs, but having proper key-value lookup felt worthwhile.

I find it a bit surprising that a map isn't a first-class feature the way arrays are (maybe there's a good reason) but this seems to be the recommended OpenAPI way to do this: https://swagger.io/docs/specification/data-models/dictionaries/

(Unless I've misinterpreted and there's something better? which is very possible)

Reply:
Ooh I see now, thank you for explaining! ok yes, this makes sense, and I think it's ok to add that stream descriptor -> string method here for now